responsemanager/server.go

```diff
@@ -229,31 +244,59 @@ func (rm *ResponseManager) processRequests(p peer.ID, requests []gsmsg.GraphSync
 			networkErrorListeners: rm.networkErrorListeners,
 			connManager:           rm.connManager,
 		}
+
+		linkSystem := result.CustomLinkSystem
+		if linkSystem.StorageReadOpener == nil {
+			linkSystem = rm.linkSystem
+		}
+
+		signals := queryexecutor.ResponseSignals{
+			PauseSignal:  make(chan struct{}, 1),
+			UpdateSignal: make(chan struct{}, 1),
+			ErrSignal:    make(chan error, 1),
+		}
+
+		responseStream := rm.responseAssembler.NewStream(rctx, p, request.ID(), subscriber)
+
+		response := &inProgressResponseStatus{
+			ctx:            rctx,
+			span:           responseSpan,
+			cancelFn:       cancelFn,
+			peer:           p,
+			request:        request,
+			linkSystem:     linkSystem,
+			customChooser:  result.CustomChooser,
+			signals:        signals,
+			startTime:      time.Now(),
+			responseStream: responseStream,
+		}
+
+		// setup query for processing
+		err := prepareQuery(rctx, p, request, result, responseStream)
+
+		if err != nil {
+			// error occurred in request hooks or setting up the query --
+			// now we're just waiting for the error response to finish sending
+			response.state = graphsync.CompletingSend
+			response.span.RecordError(err)
+			response.span.SetStatus(codes.Error, err.Error())
+		} else if result.IsPaused {
+			response.state = graphsync.Paused
+		} else {
+			response.state = graphsync.Queued
+			// TODO: Use a better work estimation metric.
+			rm.responseQueue.PushTask(p, peertask.Task{Topic: request.ID(), Priority: int(request.Priority()), Work: 1})
+		}
+
+		// save request state
 		log.Infow("graphsync request initiated", "request id", request.ID().String(), "peer", p, "root", request.Root())
 		ipr, ok := rm.inProgressResponses[request.ID()]
 		if ok && ipr.state == graphsync.Running {
 			log.Warnf("there is an identical request already in progress", "request id", request.ID().String(), "peer", p)
 		}
 
-		rm.inProgressResponses[request.ID()] =
-			&inProgressResponseStatus{
-				ctx:      rctx,
-				span:     responseSpan,
-				cancelFn: cancelFn,
-				peer:     p,
-				request:  request,
-				signals: queryexecutor.ResponseSignals{
-					PauseSignal:  make(chan struct{}, 1),
-					UpdateSignal: make(chan struct{}, 1),
-					ErrSignal:    make(chan error, 1),
-				},
-				state:          graphsync.Queued,
-				startTime:      time.Now(),
-				responseStream: rm.responseAssembler.NewStream(ctx, p, request.ID(), sub),
-			}
-		// TODO: Use a better work estimation metric.
+		rm.inProgressResponses[request.ID()] = response
```

Comment:
I don't see the path to this being cleaned up by `terminateRequest` in the case that the response is _not_ queued. Normally I think it goes into the queryexecutor and gets a `FinishTask()` from there to get cleaned up. What's the process for cleaning up `inProgressResponses` that are not registered as tasks on the `responseQueue` or are we looking at a potential leak here?

Reply:
it's not leaking but wow the thread is very unclear.

Subscriber.OnNext is called when reject message finally goes over the network (or has a network fail). That calls `responsemanager.TerminateRequest`, which calls `terminateRequest`.  https://github.com/ipfs/go-graphsync/blob/main/responsemanager/subscriber.go#L29

There needs to be some serious logic cleanup here long term as I had to refresh my own memory which is terrible.